### PR TITLE
[libc] add a simple TTAS spin lock

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
This is to be used in `pthread_spinlock`.

### CODE REVIEW

**Review by jhuber6 — COMMENTED** [review]
I've found using `fetch_or` and `fetch_and` works better on the GPU targets. That is,
```
atomic<int32_t> lock{};

bool has_lock = !lock.fetch_or(1, cpp::MemoryOrder::Acquire);
...
lock.fetch_and(0, cpp::MemoryOrder::Release);
```
Any clue how this performs on non-GPU targets?

**Comment by SchrodingerZhu** [comment]
@jhuber6 I think this technique is called CTR as in https://arxiv.org/pdf/2102.03863, section 2.1. But TTAS semantic should be more general. Frankly speaking, I am not sure how it performs on modern hardware regarding implement TTAS locks.

**Comment by SchrodingerZhu** [comment]
So according to the discussion in that paper, CTR is considered more performant in 1-to-1 communication:

>  in our case with the 1-to-1 communication protocol used on the Grant field in Hemlock, busy-waiting via
read-modify-write atomic operations provides a performance benefit. Because of the simple communication pattern, back-off in the busy-waiting loop is not useful.

**Comment by SchrodingerZhu** [comment]
@jhuber6 Updated. Please take a look.

File: libc/src/__support/threads/spin_lock.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_SRC___SUPPORT_THREADS_SPIN_LOCK_H
+#define LLVM_LIBC_SRC___SUPPORT_THREADS_SPIN_LOCK_H
+
+#include "src/__support/CPP/atomic.h"
+#include "src/__support/macros/attributes.h"
+#include "src/__support/macros/properties/architectures.h"
+#include "src/__support/threads/sleep.h"
+namespace LIBC_NAMESPACE_DECL {
```
Comment (jhuber6) [inline libc/src/__support/threads/spin_lock.h:16]:
```suggestion
#include "src/__support/threads/sleep.h"

namespace LIBC_NAMESPACE_DECL {
```

**Review by jhuber6 — APPROVED** [review]
Spin locks in general don't work on the GPU due to the lack of a fair scheduler, you need to make sure there's enough resources that any thread can make progress even if every other thread is suspended and never comes back. I'm not too concerned about the GPU implementation for this, since I don't think it'll be used in anything. My questions were mostly whether or not those bitwise operations worked better since in my experience they're considerably faster for the type of 1-to-1 waiting that's done on GPUs. However, looking at the generated assembly for x64 it's definitely not the case there. I think I'm fine with either this or the one you had before.

**Comment by SchrodingerZhu** [comment]
Yeah, spinlocks are quite useless on GPU. warp/wavefront divergence plus other scheduling issues will make a spinlock a deadlock easily. 

And for many architectures, there is no progression guarantee for cross block/CTA synchronization, so spinlock variants cannot be used across blocks/CTAs neither.

I will use the updated version anyway. Just to leave the potential for the future.

**Pull request merged**